smarts/core/bubble_manager.py

```diff
@@ -152,71 +196,81 @@ class Cursor:
     # not be the case if we spawn a vehicle in a bubble, but that wouldn't be ideal.
     vehicle: Vehicle
     state: BubbleState = None
+    transition: BubbleTransition = None
     bubble: Bubble = None
     tracking_id: str = field(init=False)
 
     def __post_init__(self):
         self.tracking_id = self.vehicle.id
 
-    def update(self) -> bool:
-        """Going through an airlock transitions a vehicle between traffic control and
-        agent control (into bubble; vice-versa out of bubble).
+    @classmethod
+    def from_pos(
+        cls, prev_pos, pos, vehicle, bubble, index, prev_cursors, running_cursors
+    ):
+        @lru_cache(maxsize=2)
+        def in_bubble(pos_):
+            return bubble.in_bubble(pos_)
+
+        @lru_cache(maxsize=2)
+        def in_airlock(pos_):
+            return bubble.in_airlock(pos_)
 
-        Returns `True` on transition (entered or exited zone)
-        """
-        pos = Point(self.vehicle.position)
-
-        state = self.state
-        if state == BubbleState.AirlockEntered and pos.within(self.bubble.geometry):
-            state = BubbleState.Entered
-        elif state is None and pos.within(self.bubble.airlock_geometry):
-            state = BubbleState.AirlockEntered
-        elif state == BubbleState.Entered and not pos.within(self.bubble.geometry):
-            state = BubbleState.Exited
-        elif state == BubbleState.Exited and not pos.within(
-            self.bubble.airlock_geometry
+        prev_pos, pos = tuple(prev_pos), tuple(pos)
+
+        is_social = vehicle.id in index.social_vehicle_ids
+        is_shadowed = index.vehicle_is_shadowed(vehicle.id)
+        is_hijacked = index.vehicle_is_hijacked(vehicle.id)
+        is_admissible = bubble.is_admissible(
+            vehicle.id, index, prev_cursors, running_cursors
+        )
+
+        transition = None
+        # TODO: Depending on step size, we could potentially skip transitions (e.g.
+        #       go straight to relinquish w/o hijacking first). This may be solved by
+        #       time-based airlocking. For robust code we'll want to handle these
+        #       scenarios (e.g. hijacking if didn't airlock first)
+        if (
+            is_social
+            and not is_shadowed
+            and is_admissible
+            and in_airlock(pos)
+            and not (in_airlock(prev_pos) or in_bubble(prev_pos))
```

Comment:
As I've been thinking of it, that's fine because they're boundaries we care about. If the bubble is travelling it may cause a transition to happen (e.g. according to these conditions here) - that's to be expected.